File: healthkit/analyzer.py
"""Cross-dimensional analysis: correlations, anomalies, trends, insights."""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd
from scipy import stats


# Phase 3 composite scores. Each takes the z-score columns produced by
# aggregator._attach_baselines() and combines them into a 0-100 daily score.
# Convention: a metric's directional contribution is +z if "higher is better"
# for that score, -z if "lower is better", -|z| if "closer to baseline is
# better". Components average; missing components are skipped (need >= 2/4).

# (zscore_column, sign) — sign = +1 / -1 / "abs" (penalize either direction)
READINESS_COMPONENTS = (
    ("hrv_zscore30",          +1),     # higher HRV = better recovery
    ("resting_hr_zscore30",   -1),     # lower RHR = better recovery
    ("sleep_score_zscore30",  +1),     # higher sleep score = better
    ("respiratory_zscore30",  "abs"),  # closer to personal baseline = better
)

ENV_STRESS_COMPONENTS = (
    ("daylight_zscore30",      -1),    # less daylight = more env stress
    ("spo2_zscore30",          -1),    # lower SpO2 = more stress
    ("respiratory_zscore30",   +1),    # higher respiratory = stress signal
    ("hrv_zscore30",           -1),    # lower HRV = more stress
)


def _composite_score(daily, components, slope: float = 20.0, midpoint: float = 50.0) -> "pd.Series":
    """Average directional z-scores → 0-100. ±2σ ≈ 90/10, ±2.5σ saturates."""
    import numpy as np
    import pandas as pd

    parts = []
    for col, sign in components:
        if col not in daily.columns:
            continue
        s = daily[col]
        if sign == "abs":
            parts.append(-s.abs())
        else:
            parts.append(s * sign)
    if not parts:
        return pd.Series(dtype=float, index=daily.index)
    stacked = pd.concat(parts, axis=1)
    # require >= 2 non-NaN components per day
    good = stacked.notna().sum(axis=1) >= 2
    z_avg = stacked.mean(axis=1, skipna=True)
    score = (midpoint + slope * z_avg).clip(0, 100)
    return score.where(good).rename(None)


def compute_readiness(daily: "pd.DataFrame") -> "pd.Series":
    """0-100 daily Readiness Score (HRV, RHR, sleep_score, respiratory)."""
    return _composite_score(daily, READINESS_COMPONENTS)


def compute_env_stress(daily: "pd.DataFrame") -> "pd.Series":
    """0-100 daily Environment Stress Score (daylight, SpO2, respiratory, HRV)."""
    return _composite_score(daily, ENV_STRESS_COMPONENTS)


METRIC_LABELS_ZH = {
    "steps": "步數",
    "active_energy": "活動消耗",
    "distance": "步行距離",
    "flights": "爬樓層",
    "hr": "平均心率",
    "resting_hr": "靜息心率",
    "walking_hr": "步行心率",
    "hrv": "HRV",
    "spo2": "血氧 (日均)",
    "spo2_sleep_min": "睡眠期間最低血氧",
    "respiratory": "呼吸頻率",
    "body_temp": "體溫",
    "vo2max": "VO2 Max",
    "sleep_hours": "睡眠時數",
    "sleep_deep_minutes": "深睡分鐘",
    "sleep_rem_minutes": "REM 分鐘",
    "sleep_awake_minutes": "夜間清醒分鐘",
    "sleep_efficiency": "睡眠效率",
    "bedtime_offset_min": "就寢時點 (18:00 後分鐘)",
    "walking_asymmetry": "步行不對稱率",
    "double_support": "雙腳支撐時間",
    "daylight": "日照時間",
    "sleep_score": "睡眠分數",
    "hr_min": "心率最低",
    "hr_max": "心率最高",
    "hr_std": "心率波動 (std)",
    "hr_samples": "心率樣本數",
    "readiness": "恢復分數",
    "env_stress": "環境壓力分數",
}


def _numeric_view(daily: pd.DataFrame) -> pd.DataFrame:
    drop_cols = [c for c in ("sleep_start", "sleep_end") if c in daily.columns]
    # Exclude derived rolling-baseline / z-score columns from correlation, anomaly,
    # and lag analyses — they're inputs to Phase 3 Readiness/EnvStress, not
    # first-class signals on their own.
    derived = [c for c in daily.columns if c.endswith("_baseline30") or c.endswith("_zscore30")]
    return daily.drop(columns=drop_cols + derived, errors="ignore").select_dtypes(include="number")


def rolling_view(daily: pd.DataFrame, window: int = 7) -> pd.DataFrame:
    num = _numeric_view(daily)
    return num.rolling(window=window, min_periods=max(2, window // 2)).mean()


def correlation_matrix(daily: pd.DataFrame, min_overlap: int = 14) -> pd.DataFrame:
    num = _numeric_view(daily)
    cols = [c for c in num.columns if num[c].notna().sum() >= min_overlap]
    if len(cols) < 2:
        return pd.DataFrame()
    return num[cols].corr(method="spearman", min_periods=min_overlap)


def lagged_correlations(
    daily: pd.DataFrame,
    drivers: list[str],
    responses: list[str],
    lags: tuple[int, ...] = (-2, -1, 0, 1, 2),
    min_overlap: int = 14,
) -> pd.DataFrame:
    """For each (driver, response, lag), compute Spearman r where response is
    shifted by `lag` days (positive lag = response measured `lag` days after driver).
    """
    num = _numeric_view(daily)
    rows = []
    for d in drivers:
        if d not in num.columns:
            continue
        for r in responses:
            if r not in num.columns or r == d:
                continue
            for lag in lags:
                shifted = num[r].shift(-lag)
                pair = pd.concat([num[d], shifted], axis=1).dropna()
                if len(pair) < min_overlap:
                    continue
                rho, p = stats.spearmanr(pair.iloc[:, 0], pair.iloc[:, 1])
                if np.isnan(rho):
                    continue
                rows.append({
                    "driver": d,
                    "response": r,
                    "lag_days": lag,
                    "spearman_r": rho,
                    "p_value": p,
                    "n": len(pair),
                })
    return pd.DataFrame(rows).sort_values(
        by=["spearman_r"], key=lambda s: s.abs(), ascending=False
    )


def detect_anomalies(daily: pd.DataFrame, sigma: float = 2.0, baseline_days: int = 28) -> pd.DataFrame:
    """Flag days where a metric deviates from its rolling baseline by >= sigma σ."""
    num = _numeric_view(daily)
    out = []
    for col in num.columns:
        s = num[col].dropna()
        if len(s) < baseline_days:
            continue
        mean = s.rolling(baseline_days, min_periods=baseline_days // 2).mean()
        std = s.rolling(baseline_days, min_periods=baseline_days // 2).std()
        z = (s - mean) / std.replace(0, np.nan)
        flagged = z.abs() >= sigma
        for date, is_flag in flagged.items():
            if is_flag and not pd.isna(z.loc[date]):
                out.append({
                    "date": date,
                    "metric": col,
                    "value": float(s.loc[date]),
                    "baseline_mean": float(mean.loc[date]),
                    "z_score": float(z.loc[date]),
                })
    if not out:
        return pd.DataFrame()
    return pd.DataFrame(out).sort_values("date", ascending=False)


@dataclass
class Insight:
    title: str
    detail: str
    severity: str  # info | watch | alert


def _label(col: str) -> str:
    return METRIC_LABELS_ZH.get(col, col)


def _trend_change(series: pd.Series, recent: int = 7, prior: int = 21) -> tuple[float, float] | None:
    """Return (recent_mean, prior_mean) if both windows are sufficiently populated."""
    s = series.dropna()
    if len(s) < recent + 5:
        return None
    recent_part = s.iloc[-recent:]
    prior_part = s.iloc[-(recent + prior):-recent] if len(s) >= recent + prior else s.iloc[:-recent]
    if len(recent_part) < max(3, recent // 2) or len(prior_part) < max(5, prior // 3):
        return None
    return float(recent_part.mean()), float(prior_part.mean())


def generate_insights(daily: pd.DataFrame) -> list[Insight]:
    """Heuristic narrative insights: trend shifts + notable correlations."""
    insights: list[Insight] = []
    num = _numeric_view(daily)

    # 1. Trend shifts on key metrics
    watch_metrics = [
        ("resting_hr", "higher_is_worse"),
        ("hrv", "higher_is_better"),
        ("sleep_hours", "higher_is_better"),
        ("sleep_efficiency", "higher_is_better"),
        ("spo2", "higher_is_better"),
        ("spo2_sleep_min", "higher_is_better"),
        ("steps", "higher_is_better"),
        ("respiratory", "neutral"),
        ("body_temp", "neutral"),
    ]
    for col, direction in watch_metrics:
        if col not in num.columns:
            continue
        change = _trend_change(num[col])
        if change is None:
            continue
        recent, prior = change
        if prior == 0 or np.isnan(prior):
            continue
        delta_pct = (recent - prior) / abs(prior) * 100
        if abs(delta_pct) < 5:
            continue

        verb = "上升" if delta_pct > 0 else "下降"
        sev = "info"
        bad = (
            (direction == "higher_is_worse" and delta_pct > 5) or
            (direction == "higher_is_better" and delta_pct < -5)
        )
        if bad:
            sev = "watch" if abs(delta_pct) < 15 else "alert"

        title = f"{_label(col)}近 7 天{verb} {abs(delta_pct):.1f}%"
        detail = f"近 7 天平均 {recent:.2f}，先前 {len(num[col].dropna()) > 28 and '21 天' or '基準期'}平均 {prior:.2f}。"
        insights.append(Insight(title=title, detail=detail, severity=sev))

    # 2. Top lagged correlation insights
    drivers = [c for c in ("sleep_hours", "sleep_efficiency", "steps", "active_energy", "bedtime_offset_min") if c in num.columns]
    responses = [c for c in ("resting_hr", "hrv", "spo2", "spo2_sleep_min", "sleep_efficiency", "sleep_hours") if c in num.columns]
    lc = lagged_correlations(daily, drivers, responses, lags=(0, 1, 2))
    if not lc.empty:
        # keep significant + meaningful effect size
        sig = lc[(lc["p_value"] < 0.05) & (lc["spearman_r"].abs() >= 0.25)]
        for _, row in sig.head(5).iterrows():
            d = _label(row["driver"])
            r = _label(row["response"])
            sign = "正相關" if row["spearman_r"] > 0 else "負相關"
            lag = int(row["lag_days"])
            lag_text = "同一天" if lag == 0 else f"延後 {lag} 天"
            insights.append(Insight(
                title=f"{d} 與 {r}（{lag_text}）呈{sign} (r={row['spearman_r']:+.2f})",
                detail=f"基於 {int(row['n'])} 天樣本，p={row['p_value']:.3f}。可作為調整生活作息的參考訊號。",
                severity="info",
            ))

    # 3. Recent anomalies (last 14 days)
    anom = detect_anomalies(daily)
    if not anom.empty:
        recent_anom = anom[anom["date"] >= (daily.index.max() - pd.Timedelta(days=14))]
        for _, row in recent_anom.head(5).iterrows():
            direction = "高於" if row["z_score"] > 0 else "低於"
            sev = "watch" if abs(row["z_score"]) < 3 else "alert"
            insights.append(Insight(
                title=f"{row['date'].strftime('%m/%d')} {_label(row['metric'])}異常 ({direction}基準 {abs(row['z_score']):.1f}σ)",
                detail=f"當日 {row['value']:.2f}，前 28 天基準平均 {row['baseline_mean']:.2f}。",
                severity=sev,
            ))

    return insights

File: docs/analyzer.js
// Cross-dimensional analysis: Spearman correlation, lagged correlation,
// rolling-baseline anomaly detection, and Chinese narrative insights.
import { columnValues } from "./aggregator.js";

export const METRIC_LABELS = {
  steps: "步數",
  active_energy: "活動消耗",
  distance: "步行距離",
  flights: "爬樓層",
  hr: "平均心率",
  resting_hr: "靜息心率",
  walking_hr: "步行心率",
  hrv: "HRV",
  spo2: "血氧 (日均)",
  spo2_sleep_min: "睡眠期間最低血氧",
  respiratory: "呼吸頻率",
  body_temp: "體溫",
  vo2max: "VO2 Max",
  sleep_hours: "睡眠時數",
  sleep_deep_minutes: "深睡分鐘",
  sleep_rem_minutes: "REM 分鐘",
  sleep_awake_minutes: "夜間清醒分鐘",
  sleep_efficiency: "睡眠效率",
  bedtime_offset_min: "就寢時點 (18:00 後分鐘)",
};

export const DRIVERS_DEFAULT = ["sleep_hours", "sleep_efficiency", "steps", "active_energy", "bedtime_offset_min"];
export const RESPONSES_DEFAULT = ["resting_hr", "hrv", "spo2", "spo2_sleep_min", "sleep_efficiency", "sleep_hours"];

export function labelOf(col) { return METRIC_LABELS[col] || col; }

// ---- statistics ----------------------------------------------------------

function rankWithTies(arr) {
  const n = arr.length;
  const idx = arr.map((v, i) => i).sort((a, b) => arr[a] - arr[b]);
  const ranks = new Array(n);
  let i = 0;
  while (i < n) {
    let j = i;
    while (j + 1 < n && arr[idx[j + 1]] === arr[idx[i]]) j++;
    const r = (i + j) / 2 + 1; // 1-based average rank
    for (let k = i; k <= j; k++) ranks[idx[k]] = r;
    i = j + 1;
  }
  return ranks;
}

function pearson(x, y) {
  const n = x.length;
  if (n < 3) return NaN;
  let mx = 0, my = 0;
  for (let i = 0; i < n; i++) { mx += x[i]; my += y[i]; }
  mx /= n; my /= n;
  let num = 0, dx = 0, dy = 0;
  for (let i = 0; i < n; i++) {
    const a = x[i] - mx, b = y[i] - my;
    num += a * b; dx += a * a; dy += b * b;
  }
  if (dx === 0 || dy === 0) return NaN;
  return num / Math.sqrt(dx * dy);
}

// erf approximation (Abramowitz & Stegun 7.1.26); good to ~1.5e-7
function erf(x) {
  const sign = x < 0 ? -1 : 1;
  const a = Math.abs(x);
  const t = 1 / (1 + 0.3275911 * a);
  const y = 1 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t + 0.254829592) * t * Math.exp(-a * a);
  return sign * y;
}

function normalCdf(z) {
  return 0.5 * (1 + erf(z / Math.SQRT2));
}

// Two-sided p-value for Spearman r using normal approximation z = r * sqrt(n-1).
// Good enough for n >= 20. Returns 1 if degenerate.
function spearmanPValue(r, n) {
  if (!Number.isFinite(r) || n < 4) return 1;
  const z = r * Math.sqrt(n - 1);
  return Math.min(1, 2 * (1 - normalCdf(Math.abs(z))));
}

export function spearman(x, y) {
  const n = x.length;
  const xs = [], ys = [];
  for (let i = 0; i < n; i++) {
    if (Number.isFinite(x[i]) && Number.isFinite(y[i])) {
      xs.push(x[i]); ys.push(y[i]);
    }
  }
  if (xs.length < 4) return { r: NaN, p: 1, n: xs.length };
  const r = pearson(rankWithTies(xs), rankWithTies(ys));
  return { r, p: spearmanPValue(r, xs.length), n: xs.length };
}

// ---- analysis primitives -------------------------------------------------

export function analyzableColumns(frame, minObs = 14) {
  return frame.columns.filter((c) => {
    const vals = columnValues(frame, c);
    let n = 0;
    for (const v of vals) if (Number.isFinite(v)) n++;
    return n >= minObs;
  });
}

export function correlationMatrix(frame, cols) {
  const data = {};
  for (const c of cols) data[c] = columnValues(frame, c);
  const matrix = cols.map(() => new Array(cols.length).fill(NaN));
  for (let i = 0; i < cols.length; i++) {
    for (let j = 0; j < cols.length; j++) {
      if (i === j) { matrix[i][j] = 1; continue; }
      if (j < i) { matrix[i][j] = matrix[j][i]; continue; }
      matrix[i][j] = spearman(data[cols[i]], data[cols[j]]).r;
    }
  }
  return { cols, matrix };
}

function shifted(arr, lag) {
  // shift response by `lag`: positive lag = response measured `lag` days after driver
  // we align driver[i] with response[i + lag]
  const out = new Array(arr.length).fill(NaN);
  for (let i = 0; i < arr.length; i++) {
    const j = i + lag;
    out[i] = (j >= 0 && j < arr.length) ? arr[j] : NaN;
  }
  return out;
}

export function laggedCorrelations(frame, drivers, responses, lags, minObs = 14) {
  const out = [];
  for (const d of drivers) {
    if (!frame.columns.includes(d)) continue;
    const dx = columnValues(frame, d);
    for (const r of responses) {
      if (!frame.columns.includes(r) || r === d) continue;
      const ry = columnValues(frame, r);
      for (const lag of lags) {
        const shifted_y = shifted(ry, lag);
        const { r: rho, p, n } = spearman(dx, shifted_y);
        if (!Number.isFinite(rho) || n < minObs) continue;
        out.push({ driver: d, response: r, lag, r: rho, p, n });
      }
    }
  }
  out.sort((a, b) => Math.abs(b.r) - Math.abs(a.r));
  return out;
}

export function detectAnomalies(frame, sigma = 2.0, baseline = 28) {
  const anomalies = [];
  for (const col of frame.columns) {
    const vals = columnValues(frame, col);
    // rolling mean/std using only finite values within the trailing window
    for (let i = baseline; i < vals.length; i++) {
      const v = vals[i];
      if (!Number.isFinite(v)) continue;
      let n = 0, s = 0, ss = 0;
      for (let j = Math.max(0, i - baseline); j < i; j++) {
        const u = vals[j];
        if (Number.isFinite(u)) { n++; s += u; ss += u * u; }
      }
      if (n < Math.floor(baseline / 2)) continue;
      const mean = s / n;
      const variance = Math.max(0, ss / n - mean * mean);
      const sd = Math.sqrt(variance);
      if (sd === 0) continue;
      const z = (v - mean) / sd;
      if (Math.abs(z) >= sigma) {
        anomalies.push({ date: frame.rows[i].date, metric: col, value: v, baseline_mean: mean, z });
      }
    }
  }
  anomalies.sort((a, b) => (a.date < b.date ? 1 : -1));
  return anomalies;
}

// ---- insights ------------------------------------------------------------

function meanFinite(arr) {
  let n = 0, s = 0;
  for (const v of arr) if (Number.isFinite(v)) { n++; s += v; }
  return n ? s / n : NaN;
}

function trendChange(values, recent = 7, prior = 21) {
  // Find last `recent` finite samples and the prior `prior` samples before that
  const finite = [];
  for (let i = 0; i < values.length; i++) if (Number.isFinite(values[i])) finite.push(values[i]);
  if (finite.length < recent + 5) return null;
  const recentArr = finite.slice(-recent);
  const start = Math.max(0, finite.length - recent - prior);
  const priorArr = finite.slice(start, finite.length - recent);
  if (recentArr.length < Math.max(3, Math.floor(recent / 2))) return null;
  if (priorArr.length < Math.max(5, Math.floor(prior / 3))) return null;
  return { recent: meanFinite(recentArr), prior: meanFinite(priorArr) };
}

const WATCH = [
  ["resting_hr",       "higher_is_worse"],
  ["hrv",              "higher_is_better"],
  ["sleep_hours",      "higher_is_better"],
  ["sleep_efficiency", "higher_is_better"],
  ["spo2",             "higher_is_better"],
  ["spo2_sleep_min",   "higher_is_better"],
  ["steps",            "higher_is_better"],
  ["respiratory",      "neutral"],
  ["body_temp",        "neutral"],
];

export function generateInsights(frame) {
  const insights = [];
  // 1. Trend shifts
  for (const [col, dir] of WATCH) {
    if (!frame.columns.includes(col)) continue;
    const ch = trendChange(columnValues(frame, col));
    if (!ch || !Number.isFinite(ch.prior) || ch.prior === 0) continue;
    const deltaPct = (ch.recent - ch.prior) / Math.abs(ch.prior) * 100;
    if (Math.abs(deltaPct) < 5) continue;
    const verb = deltaPct > 0 ? "上升" : "下降";
    let severity = "info";
    const bad = (dir === "higher_is_worse" && deltaPct > 5) ||
                (dir === "higher_is_better" && deltaPct < -5);
    if (bad) severity = Math.abs(deltaPct) < 15 ? "watch" : "alert";
    insights.push({
      severity,
      title: `${labelOf(col)}近 7 天${verb} ${Math.abs(deltaPct).toFixed(1)}%`,
      detail: `近 7 天平均 ${ch.recent.toFixed(2)}，先前 21 天平均 ${ch.prior.toFixed(2)}。`,
    });
  }

  // 2. Top significant lagged correlations
  const drivers = DRIVERS_DEFAULT.filter((c) => frame.columns.includes(c));
  const responses = RESPONSES_DEFAULT.filter((c) => frame.columns.includes(c));
  const lc = laggedCorrelations(frame, drivers, responses, [0, 1, 2]);
  const significant = lc.filter((x) => x.p < 0.05 && Math.abs(x.r) >= 0.25);
  for (const row of significant.slice(0, 5)) {
    const lagText = row.lag === 0 ? "同一天" : `延後 ${row.lag} 天`;
    const sign = row.r > 0 ? "正相關" : "負相關";
    insights.push({
      severity: "info",
      title: `${labelOf(row.driver)} 與 ${labelOf(row.response)}（${lagText}）呈${sign} (r=${row.r >= 0 ? "+" : ""}${row.r.toFixed(2)})`,
      detail: `基於 ${row.n} 天樣本，p=${row.p.toFixed(3)}。可作為調整生活作息的參考訊號。`,
    });
  }

  // 3. Recent anomalies (last 14 days)
  const anom = detectAnomalies(frame);
  if (anom.length && frame.rows.length) {
    const lastDate = new Date(frame.rows[frame.rows.length - 1].date);
    const cutoff = new Date(lastDate);
    cutoff.setDate(cutoff.getDate() - 14);
    const recent = anom.filter((a) => new Date(a.date) >= cutoff).slice(0, 5);
    for (const a of recent) {
      const dir = a.z > 0 ? "高於" : "低於";
      const severity = Math.abs(a.z) < 3 ? "watch" : "alert";
      const md = a.date.slice(5);
      insights.push({
        severity,
        title: `${md} ${labelOf(a.metric)}異常 (${dir}基準 ${Math.abs(a.z).toFixed(1)}σ)`,
        detail: `當日 ${a.value.toFixed(2)}，前 28 天基準平均 ${a.baseline_mean.toFixed(2)}。`,
      });
    }
  }

  return insights;
}
